### PR TITLE
Add in ML by Target and redo ML classes to be schema

### LINKED ISSUE #425: ML models need to include target information and be spec'd properly by date
ML models need to include target information and be spec'd properly by date so we can pull the most recent ones down for our workflow.

### PR DESCRIPTION
**Pull request by hmacdope**
Fixes #425

### CODE REVIEW

**Comment by hmacdope** [comment]
@jthorton @kaminow Could I grab a review here? 

This is a big re-work of the way models are spec'd pulled and can be grabbed in a workflow all done with lots of schema.
 
1. The YAML file now has a schema `MLModelRegistry` made of subschema `MLModelSpec` which you can dynamically pull models from by date and target. 
2. Some changes to the YAML files themselves
3. The new Mac1 model is ready deployed and will be auto selected in the e2e WF
4. The models have a schema for when they are pulled down `LocalMLModelSpec`, making them easy to instantiate from local files
5. The inference classes themselves are schema. I still need to work out how to serialize the `mtenn` pytorch models or possibly even drop them and then reconstruct but I think I want to push this to a seperate PR.

File: asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+            weights_file=Path(weights_file),
+            config_file=Path(config_file) if self.config_resource else None,
+            last_updated=self.last_updated,
+            targets=self.targets,
+            local_dir=Path(local_dir) if local_dir else None,
+            build_model_kwargs=self.build_model_kwargs,
+        )
+
+
+class LocalMLModelSpec(MLModelBase):
```
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py:125]: actually, on second thought, I'm not sure I see a reason to have both `MLModelSpec` and `LocalMLModelSpec`. I think we could probably unify them into one, with a validator to make sure that there are either URLs+hashes, or local paths. in this case, `pull()` would populate the local path fields (and potentially load the weights/config into fields as well). it's also possible I'm just missing something though and both of these are useful to have separately.

**Review by kaminow — CHANGES_REQUESTED** [review]
love these new features! this will definitely help keep things more organized. I had a couple questions/thoughts, but I think the main one is regarding the necessity of having both the `MLModelSpec` and `LocalMLModelSpec` schema

**Comment by kaminow** [comment]
> The inference classes themselves are schema. I still need to work out how to serialize the mtenn pytorch models or possibly even drop them and then reconstruct but I think I want to push this to a seperate PR.

regarding this, as long as the config used to construct the model is tracked along with the model weights, it should be trivial to reconstruct the `mtenn` models

File: asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+        Returns
+        -------
+        List[str]
+            List of valid model types
+        """
+        return [model_type.value for model_type in cls]
+
+
+_SPECIAL_BUILD_MODEL_KWARGS = {MLModelType.schnet: {"pred_r": "pIC50"}}
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py:42]: Ok that makes sense, let me check how big a change this is.

File: asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+            raise FileNotFoundError(f"Yaml spec file {yaml_file} does not exist")
+
+        with open(yaml_file) as f:
+            spec = yaml.safe_load(f)
+
+        models = {}
+        for model in spec:
+            model_data = spec[model]
+            has_config = "config" in model_data
+            models[model] = MLModelSpec(
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py:263]: I'll add 

```
Config:
   validate_assignment = True

```

which forces pydantic to do it.

File: asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py:125]: This makes sense, might reduce some complexity, ill investigate.

File: asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py]: Ok fixed, great suggestion.

File: asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py:121]: I think I might leave this the way it is for a few reasons.

1. Making MLModelSpec dual purpose means you will need to check all the time if a model has been pulled and what the status of the fields are. If MLModelSpec is able to hold the local or remote spec basically every field becomes optional. 

2. LocalMLModel spec can be used to easily instantiate an inference class from some local weights lying around without having to specify all the other URL stuff say for local inference, testing and training if needed. 

3. There is a clearer mapping to the YAML file, MLModelSpec is the direct schema for the file.

[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models/ml_models.py:121]: We can discuss further at our meeting tommorow and also we can of course change in the future. :)

**Review by kaminow — APPROVED** [review]
looks great!

**Pull request merged**